components/scream/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -0,0 +1,285 @@
+        aer_tau_sw_src_sub(k) = linear_interp(aer_tau_sw_beg_sub(k), aer_tau_sw_end_sub(k), t_norm);
+      });
+    });
+    team.team_barrier();
+    }
+    {
+    /* Loop over all LW variables with nlwbands */
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nlwbands), [&] (int n) {
+      const auto& aer_tau_lw_beg_sub         = ekat::subview(data_beg.AER_TAU_LW, i, n);
+      const auto& aer_tau_lw_end_sub         = ekat::subview(data_end.AER_TAU_LW, i, n);
+      const auto& aer_tau_lw_src_sub         = ekat::subview(aer_tau_lw_src, i, n);
+
+      /* Now loop over fastest index, the number of vertical packs */
+      Kokkos::parallel_for(
+        Kokkos::ThreadVectorRange(team, nk_pack), [&] (Int k) {
+        aer_tau_lw_src_sub(k) = linear_interp(aer_tau_lw_beg_sub(k), aer_tau_lw_end_sub(k), t_norm);
+      });
+    });
+    team.team_barrier();
+    }
+  });
+  Kokkos::fence();
+
+  // Third Step: Vertical interpolation, project the SPA data onto the pressure profile for this simulation.
+  // This is done using the EKAT linear interpolation routine, see /externals/ekat/util/ekat_lin_interp.hpp
+  // for more details. 
+  using LIV = ekat::LinInterp<Real,Spack::n>;
+  Real minthreshold = 0.0;  // Hard-code a minimum value for aerosol concentration to zero.
+
+  LIV VertInterp(ncols_atm,pressure_state.nlevs,nlevs_atm,minthreshold);
+  /* Parallel loop strategy:
+   * 1. Loop over all simulation columns (i index)
+   * 2. Where applicable, loop over all aerosol bands (n index)
+   */ 
+  Kokkos::parallel_for("vertical-interp-spa",
+    VertInterp.m_policy,
+    KOKKOS_LAMBDA(typename LIV::MemberType const& team) {
+    const int i = team.league_rank();
+    /* Setup the linear interpolater for this column. */
+    VertInterp.setup(team,
+                     ekat::subview(p_src,i),
+                     ekat::subview(pressure_state.pmid,i));
+    team.team_barrier();
+    /* Conduct vertical interpolation for the 2D variable CCN3 */
+    VertInterp.lin_interp(team,
+                          ekat::subview(p_src,i),
+                          ekat::subview(pressure_state.pmid,i),
+                          ekat::subview(ccn3_src,i),
+                          ekat::subview(data_out.CCN3,i));
+    /* Conduct vertical interpolation for the LW banded data - nlwbands (n index) */
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nlwbands), [&] (int n) {
+      VertInterp.lin_interp(team,
+                            ekat::subview(p_src,i),
+                            ekat::subview(pressure_state.pmid,i),
+                            ekat::subview(aer_tau_lw_src,i,n),
+                            ekat::subview(data_out.AER_TAU_LW,i,n));
+    });
+    /* Conduct vertical interpolation for the SW banded data - nswbands (n index) */
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nswbands), [&] (int n) {
+      VertInterp.lin_interp(team,
```

Comment:
You'll need to set up a ThreadVectorRange here:
```
const auto& tvr = Kokkos::ThreadVectorRange(team, VertInterp.km2_pack());
```
... and pass it as the second arg to lin_interp.

Reply:
@jgfouca , Does this mean that all `lin_interp` calls need an extra argument now?

Reply:
@AaronDonahue , just when you are using a custom boundary, so the lin_interp calls that occur inside a TeamThreadRange kernel.

Reply:
@AaronDonahue , I gave you bad code (doesn't work on CUDA). You'll need to capture `VertInterp.km2_pack()` in a constant before the kernel is launched. You can do this right after the VertInterp object is constructed.

Reply:
@bartgol  said there may be a fix coming in EKAT.  Should I just wait on that?  Or would I still need to capture VertInterp.km2_pack() as a constant?

Reply:
I would wait for the fix. It will allow a cleaner code, without the need for the local temporary.

Reply:
I wanted to follow up on this since the comment is a bit old.  Is there anything left to do here?  Should I be deleting the `tvr` variables being passed to the `lin_interp` calls?

Reply:
No, no, you need to keep those `tvr`'s.